sdk/go/common/util/cmdutil/child_windows.go

```diff
@@ -11,65 +11,122 @@
+// longer live.
+//
+// `KillChildren` is Intended to be used with `RegisterProcessGroup`
+// to make sure misbehaving commands do not leave any orphan
+// sub-processes running:
+//
+//	cmd := exec.Command(name, args...)
+//	cmdutil.RegisterProcessGroup(cmd)
+//      cmd.Start() // or any other method that actually starts the process
+//      err := cmdutil.KillChildren(cmd.Process.Pid)
+//
+// This is the Windows-specific implementation that scans all system
+// processes using native syscalls (via go-ps) and attempts to kill
+// them, aggregating any errors it encounters.
 func KillChildren(pid int) error {
 	procs, err := ps.Processes()
 	if err != nil {
 		return err
 	}
 
-	var result error
+	descendants := filterDescendants(pid, procs)
 
-	for _, proc := range procs {
-		if proc.PPid() == pid {
-			toKill, err := os.FindProcess(proc.Pid())
-			if err != nil {
-				// It's possible that the process has already exited, let's see if it still exists. Either way, we won't
-				// try to kill it but we will add the original error from os.FindProcess() if we can't prove it doesn't
-				// exits.
-				exists, existsErr := processExistsWithParent(proc.Pid(), proc.PPid())
-				if existsErr != nil || exists {
-					result = multierror.Append(result, err)
-				}
-				continue
-			}
+	// Try to kill the descendants and collect errors by PID.
+	// These errors may not be relevant if the descendant
+	// terminates in some other way.
+	errors := map[int]error{}
+	for _, proc := range descendants {
+		procPid := proc.Pid()
+		toKill, err := os.FindProcess(procPid)
+		if err != nil {
+			errors[procPid] = fmt.Errorf("FindProcess(%d) failed: %w", procPid, err)
+		}
+
+		err = toKill.Kill()
+		if err != nil {
+			errors[procPid] = fmt.Errorf("proc.Kill() failed for pid=%d: %w", procPid, err)
+		}
+	}
+
+	survivingProcesses, err := ps.Processes()
+	if err != nil {
+		return err
+	}
 
-			err = toKill.Kill()
-			if err != nil {
+	survivingPids := map[int]bool{}
+	for _, p := range survivingProcesses {
+		survivingPids[p.Pid()] = true
+	}
+
+	// Only report errors for descendants that survived our
+	// attempt to kill them.
+	//
+	// There are races inherent in sending a `Kill()` and
+	// observing the process in the `survivingPids`. This method
+	// would rather return a `nil` error when unsure than error
+	// out incorrectly.
+	var result error
+
+	for _, proc := range descendants {
+		_, surviving := survivingPids[proc.Pid()]
+		if surviving {
+			err, gotErr := errors[proc.Pid()]
+			if gotErr {
 				result = multierror.Append(result, err)
 			}
 		}
 	}
-
 	return result
 }
 
-func processExistsWithParent(pid int, ppid int) (bool, error) {
-	procs, err := ps.Processes()
-	if err != nil {
-		return false, err
+func filterDescendants(rootPid int, processes []ps.Process) []ps.Process {
+	parents := map[int]int{}
+	for _, p := range processes {
+		pid := p.Pid()
+		ppid := p.PPid()
+		// Can have PID=0 with PPID=0, ignore it.
```

Comment:
Is 0 special here or do you just mean that when PID and PPID are equal to ignore them?

Reply:
Experimentally it seems that the 0 is the special PID but I'm not sure. What I care about is there's no cycles in this graph. x.PPID=x.PID creates a cycle that can make the program below loop.